optuna/_deprecated.py

```diff
@@ -0,0 +1,138 @@
+
+
+_DEPRECATION_NOTE_TEMPLATE = """
+
+.. note::
+    Deprecated in v{d_ver}. This feature will be removed in the future. The removal of this
+    feature is currently scheduled for v{r_ver}, but this schedule is subject to change.
+    See https://github.com/optuna/optuna/releases/tag/v{d_ver}.
+"""
+
+
+def _validate_two_version(old_version: str, new_version: str) -> None:
+    if version.parse(old_version) > version.parse(new_version):
+        raise ValueError(
+            "Invalid version relationship. The deprecated version must be smaller than "
+            "the removed version, but (deprecated version, removed version) = ({}, {}) are "
+            "specified.".format(old_version, new_version)
+        )
+
+
+def _formatting_note(note: str) -> str:
+    return "\n\n.. note::\n" + "".join(["    " + s + "\n" for s in note.split(sep="\n")])
+
+
+def deprecated(
+    deprecated_version: str, removed_version: str, name: str = None, deprecation_note: str = None
+) -> Any:
+    """Decorate class or function as deprecated.
+
+    Args:
+        deprecated_version: The version in which the target feature is deprecated.
+        removed_version: The version in which the target feature will be removed.
+        name: The name of the feature. Defaults to the function or class name. Optional.
+        deprecation_note:
+            The text for the deprecation note. The default note is build using specified
+            ``deprecated_version`` and ``removed_version``. If you want to provide additional
+            information, please specify this argument yourself.
+
+            .. note::
+                The default deprecation note is as follows: "Deprecated in v{d_ver}. This feature
+                will be removed in the future. The removal of this feature is currently scheduled
+                for v{r_ver}, but this schedule is subject to change. See
+                https://github.com/optuna/optuna/releases/tag/v{d_ver}."
+
+            .. note::
+                The specified text must be separated by the newline characters, and each line
+                should not contain more than 92 characters without the added newline character.
+    """
+
+    _validate_version(deprecated_version)
+    _validate_version(removed_version)
+    _validate_two_version(deprecated_version, removed_version)
+
+    def _deprecated_wrapper(f: Any) -> Any:
+        # f is either func or class.
+
+        def _deprecated_func(func: Callable[[Any], Any]) -> Callable[[Any], Any]:
+            """Decorates a function as deprecated.
+
+            This decorator is supposed to be applied to the deprecated function.
+            """
+            if func.__doc__ is None:
+                func.__doc__ = ""
+
+            note = (
+                _DEPRECATION_NOTE_TEMPLATE.format(d_ver=deprecated_version, r_ver=removed_version)
+                if deprecation_note is None
+                else _formatting_note(deprecation_note)
```

Comment:
Is it difficult to append the `deprecation_note` to the `_DEPRECATION_NOTE_TEMPLATE`, or do you prefer this way using either? I think it's simpler always keeping the template, thinking of this note as an optional body after it.